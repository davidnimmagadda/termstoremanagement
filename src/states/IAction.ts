/**
 * We are following FSA (Flux Standard Action) compliant actions, which is mentioned in Redux documentations.
 * This Flux is not a typo :)
 */
export interface IAction {
  /**
   * We use strings to represent action types.
   * Each page should have separated definitions for all allowed types.
   *
   * For example:
   *
   * type enum ActionTypes = 'ADD_ITEM' | 'REMOVE_ITEM';
   */
  type: string;

  /**
   * The optional payload can be anything holds, well, the payload.
   * If `error` is `true`, the payload should be an error object.
   */
  // tslint:disable-next-line:no-any
  payload?: any;

  /**
   * A `true` value means the action has an error.
   * Any other values, including `undefined` and `null`, means the action is good.
   */
  error?: boolean;

  /**
   * The optional meta can be anything holds extra information that is not part of the payload.
   */
  // tslint:disable-next-line:no-any
  meta?: any;
}
