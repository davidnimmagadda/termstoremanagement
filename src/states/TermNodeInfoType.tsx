/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

enum TermNodeInfoType {
  FOLDER = 'folder',
  FILE = 'file'
}

export default TermNodeInfoType;
