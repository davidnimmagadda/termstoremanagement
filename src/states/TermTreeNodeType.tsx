/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

enum TermTreeNodeType {
  EMPTY = '',
  TERM_STORE = 'TERM_STORE',
  TERM_GROUP = 'TERM_GROUP',
  TERM_SET = 'TERM_SET',
  TERM = 'TERM'
}

export default TermTreeNodeType;
