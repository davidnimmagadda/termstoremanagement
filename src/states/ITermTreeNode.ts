/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

import TermTreeNodeType from './TermTreeNodeType';
import TermNodeInfoType from './TermNodeInfoType';

export interface ITermTreeNode {
  nodeType: TermTreeNodeType;
  level: number;
  childrenUri: string;
  isOpen?: boolean;
  info: ITermTreeNodeAPIResponse;
}

export interface ITermTreeNodeAPIResponse {
  id: string;
  type: TermNodeInfoType;
  name: string;
}
