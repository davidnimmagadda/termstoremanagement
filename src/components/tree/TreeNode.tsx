/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

import * as React from "react";
import { Spinner, Icon, IconButton } from "office-ui-fabric-react";

import { ITermTreeNode } from "../../states/ITermTreeNode";
import "../../TermStoreAdminCenterPage.css";
import TermNodeInfoType from "../../states/TermNodeInfoType";

function getPaddingLeft(level: number, type: TermNodeInfoType): number {
  return level * 20;
}

export interface ITreeNodeProps {
  currentNode: ITermTreeNode;
  show: boolean;
  selectedTermTreeNode: ITermTreeNode;
  onNodeSelect: (node: ITermTreeNode) => void;
  loadNodeChildren: (currentNode: ITermTreeNode) => Promise<ITermTreeNode[]>;
}

export function TreeNode(props: ITreeNodeProps): JSX.Element {
  const [node, setNode] = React.useState(props.currentNode);
  const [children, setChildren] = React.useState([] as ITermTreeNode[]);
  const [loading, setLoading] = React.useState(false);
  const [highlighted, setHighLighted] = React.useState(false);

  React.useEffect(() => {
    if (
      props.selectedTermTreeNode &&
      props.selectedTermTreeNode.nodeType === node.nodeType &&
      props.selectedTermTreeNode.info.id === node.info.id
    ) {
      setNode(props.selectedTermTreeNode);
    }
  }, [props.selectedTermTreeNode]);

  function onToggle(): void {
    if (!node.isOpen) {
      // call load children
      loadChildren();
    }
    setNode({ ...node, isOpen: !node.isOpen });
  }

  async function loadChildren(): Promise<void> {
    if (
      children.length === 0 &&
      node.info &&
      node.info.type === TermNodeInfoType.FOLDER
    ) {
      setLoading(true);
      try {
        const response: ITermTreeNode[] = await props.loadNodeChildren(node);
        setChildren(response);
      } catch (err) {
        // TODO danimmag: 'error in api call TODO: handle';
      }
      setLoading(false);
    }
  }

  function getChevron(): string {
    return node.info.type === TermNodeInfoType.FOLDER && node.isOpen
      ? "ChevronDownMed"
      : "ChevronRightMed";
  }

  function getFolderIcon(): string {
    if (node.info.type === TermNodeInfoType.FILE) {
      return "Script";
    } else if (node.info.type === TermNodeInfoType.FOLDER) {
      return node.isOpen ? "FabricOpenFolderHorizontal" : "FabricFolderFill";
    } else {
      return "";
    }
  }

  function enableHighlight(): void {
    setHighLighted(true);
  }

  function disableHighlight(): void {
    setHighLighted(false);
  }

  return (
    <>
      <div
        className="treeNode"
        style={{
          paddingLeft: getPaddingLeft(node.level, node.info.type),
          display: props.show ? "flex" : "none"
        }}
        onMouseEnter={enableHighlight}
        onMouseLeave={disableHighlight}
      >
        {loading ? (
          <Spinner className="treeNodeIcon" />
        ) : (
          <span
            role="button"
            data-automation-id="treeNodeToggleChevron"
            onClick={onToggle}
          >
            <Icon className="treeNodeIcon" iconName={getChevron()} />
          </span>
        )}
        <Icon style={{ marginRight: 5 }} iconName={getFolderIcon()} />
        <span
          role="button"
          data-automation-id="treeNodeDisplayName"
          style={{ whiteSpace: "nowrap", width: "100%" }}
          onClick={() => props.onNodeSelect(node)}
        >
          {node.info.name}
        </span>
        <IconButton
          style={{ visibility: highlighted ? "initial" : "hidden" }}
          iconProps={{ iconName: "MoreVertical" }}
          onClick={() => alert("I'm Clicked!")} // TODO danimmag: to be implemented
        />
      </div>

      {children.map(childNode => (
        <TreeNode
          key={childNode.info.id}
          show={props.show && node.isOpen ? true : false}
          currentNode={childNode}
          onNodeSelect={props.onNodeSelect}
          selectedTermTreeNode={props.selectedTermTreeNode}
          loadNodeChildren={props.loadNodeChildren}
        />
      ))}
    </>
  );
}

export default TreeNode;
