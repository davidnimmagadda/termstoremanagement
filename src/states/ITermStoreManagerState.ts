/**
 * @copyright Microsoft Corporation. All rights reserved.
 */
import { ITermTreeNode } from './ITermTreeNode';

export interface ITermStoreManagerState {
  selectedTermTreeNode: ITermTreeNode;
}
