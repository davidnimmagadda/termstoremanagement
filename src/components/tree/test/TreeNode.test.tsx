/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";

import TreeNode, { ITreeNodeProps } from "../TreeNode";
import TermTreeNodeType from "../../../states/TermTreeNodeType";
import TermNodeInfoType from "../../../states/TermNodeInfoType";
import { ITermTreeNode } from "../../../states/ITermTreeNode";

describe("Term Store Admin Center Page", () => {
  describe("Tree Node", () => {
    const node: ITermTreeNode = {
      nodeType: TermTreeNodeType.TERM_STORE,
      level: 0,
      childrenUri: "termGroups",
      info: {
        id: "abc",
        type: TermNodeInfoType.FOLDER,
        name: "Taxonomy term store"
      }
    };
    const treeNodeProps: ITreeNodeProps = {
      currentNode: node,
      show: true,
      selectedTermTreeNode: node,
      onNodeSelect: jest.fn(),
      loadNodeChildren: jest.fn()
    };

    it("should render same as previous", () => {
      const tree: ReactTestRenderer = create(<TreeNode {...treeNodeProps} />);
      expect(tree).toMatchSnapshot();
    });

    it("should call load children on chevron click", () => {
      const component: ShallowWrapper = shallow(
        <TreeNode {...treeNodeProps} />
      );
      component
        .find('[data-automation-id="treeNodeToggleChevron"]')
        .simulate("click");
      expect(treeNodeProps.loadNodeChildren).toHaveBeenCalled();
    });

    it("should call load 2 children on chevron click", () => {
      (treeNodeProps.loadNodeChildren as jest.Mock).mockReturnValue(
        Promise.resolve([
          { id: "12", type: TermNodeInfoType.FOLDER, name: "cdscs csd" },
          { id: "13", type: TermNodeInfoType.FOLDER, name: "dsdc csd" }
        ])
      );
      const component: ShallowWrapper = shallow(
        <TreeNode {...treeNodeProps} />
      );
      component
        .find('[data-automation-id="treeNodeToggleChevron"]')
        .simulate("click");
      expect(treeNodeProps.loadNodeChildren).toHaveBeenCalledWith(
        treeNodeProps.currentNode
      );
    });

    it("should call on node select on click of node display name", () => {
      const component: ShallowWrapper = shallow(
        <TreeNode {...treeNodeProps} />
      );
      component
        .find('[data-automation-id="treeNodeDisplayName"]')
        .simulate("click");
      expect(treeNodeProps.onNodeSelect).toHaveBeenCalled();
    });
  });
});
