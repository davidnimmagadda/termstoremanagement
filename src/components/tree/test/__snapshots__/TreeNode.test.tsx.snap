// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Term Store Admin Center Page Tree Node should render same as previous 1`] = `
<div
  className="treeNode"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "display": "flex",
      "paddingLeft": 0,
    }
  }
>
  <span
    data-automation-id="treeNodeToggleChevron"
    onClick={[Function]}
    role="button"
  >
    <i
      aria-hidden={true}
      className="treeNodeIcon root-47"
      data-icon-name="ChevronRightMed"
      role="presentation"
    >
      
    </i>
  </span>
  <i
    aria-hidden={true}
    className="root-49"
    data-icon-name="FabricFolderFill"
    role="presentation"
    style={
      Object {
        "marginRight": 5,
      }
    }
  >
    
  </i>
  <span
    data-automation-id="treeNodeDisplayName"
    onClick={[Function]}
    role="button"
    style={
      Object {
        "whiteSpace": "nowrap",
        "width": "100%",
      }
    }
  >
    Taxonomy term store
  </span>
  <button
    className="ms-Button ms-Button--icon root-50"
    data-is-focusable={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
    type="button"
  >
    <div
      className="ms-Button-flexContainer flexContainer-51"
    >
      <i
        aria-hidden={true}
        className="ms-Button-icon icon-59"
        data-icon-name="MoreVertical"
        role="presentation"
      >
        
      </i>
    </div>
  </button>
</div>
`;
